### PR TITLE
Domain: Refactor to use a common domain selection view

### LINKED ISSUE #22246: Domain Selection: Refactor to use a common domain selection view
Refactor to use a new domain selection that was created for site creation

| Domains in site creation                                           | Domains in other places                                            |
|---------------------------------------------------------------------|-------------------------------------------------------------------|
| Image | Image |

### PR DESCRIPTION
**Pull request by staskus**
Fixes #22246

Recently we created a new modern domain selection for Site Creation flow, but we continued using the previously created domain selection in all the other flows. The views not only differ in UI but also in how many domains we fetch and how we sort them. 

The goal of this PR is to have a single implementation of domain selection for all the flows.

### Solution

I first identified different entry points and their requirements:

| Name | Domain selection | Flow |
| -------- | -------- | -------- |
|  Free to Paid Dashboard Card         |  Paid domains with free for the first year note        |    Card -> Domains -> Plans -> Checkout      |
|  Register Dashboard Card        |  Paid domains with free for the first year note   | Card -> Domains -> Registration form  |
|  All Domains        |  Paid domains with free for the first year note and transfer domain note       |  Me -> All Domains -> + -> Domains -> Just buy a domain -> Checkout   /  Me -> All Domains -> + -> Domains -> Existing WordPress.com site -> Choose Site -> Plans -> Checkout     |
|  Site Domains        |   Paid domains with site redirect note       |    Dashboard -> More -> Domains -> Search for a domain -> Domains -> Checkout      |

Then turned`WebAddressWizardContent` into `DomainSelectionViewController` and started moving functionality one by one from `RegisterDomainSuggestionsViewController` to `DomainSelectionViewController` so it would support all the cases for the site creation as well as all the previous functionality. I didn't want to change anything outside domain selection view controllers and keep everything else working in same way.

In the end `DomainSelectionViewController` grew and now includes the business logic within it. One step I haven't done is move away `switch` statements into different configuration files using composition. For now, I feel it's more convenient having all the logic in one place.

## Testing

- Test existing entry points to open a new domain selection flow
- Confirm that previously created flows continue to be working as before

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/8653a752-1541-44c1-ab33-b3608ee13791

## Regression Notes
1. Potential unintended areas of impact

Breaking domain selection in all the previous flows.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift [hunk 1]
```diff
@@ -97,12 +115,46 @@ final class WebAddressWizardContent: CollapsableHeaderViewController {
         }
     }
 
+    // MARK: - Transfer Footer Views
```
Comment (staskus) [inline WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift:110]: Copied from `RegisterDomainSuggestionsViewController`

File: WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift [hunk 2]
```diff
@@ -326,8 +393,52 @@ final class WebAddressWizardContent: CollapsableHeaderViewController {
...
+
         trackDomainsSelection(selectedDomain)
-        selection(selectedDomain)
+
+        let onFailure: () -> () = { [weak self] in
+            self?.setPrimaryButtonLoading(false, afterDelay: 0.25)
+            self?.displayActionableNotice(title: Strings.errorTitle, actionTitle: Strings.errorDismiss)
+        }
+
+        switch domainSelectionType {
```
Comment (staskus) [inline WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift:373]: Copied from `RegisterDomainSuggestionsViewController`

File: WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift [hunk 3]
```diff
@@ -715,3 +844,198 @@ extension WebAddressWizardContent: UITableViewDelegate {
         performSearchIfNeeded(query: retryQuery)
     }
 }
+
+// MARK: - Transfer Footer Setup
```
Comment (staskus) [inline WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift:780]: Copied from `RegisterDomainSuggestionsViewController`

File: WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift [hunk 4]
```diff
@@ -715,3 +844,198 @@ extension WebAddressWizardContent: UITableViewDelegate {
...
+
+    private func hideTransferFooterView(animated: Bool = true) {
+        self.updateTransferFooterViewConstraints(hidden: true, animated: animated)
+    }
+}
+
+// MARK: - Support
+
+private extension DomainSelectionViewController {
+    func includeSupportButtonIfNeeded() {
```
Comment (staskus) [inline WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift:827]: Copied from `RegisterDomainSuggestionsViewController`

File: WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift [hunk 5]
```diff
@@ -715,3 +844,198 @@ extension WebAddressWizardContent: UITableViewDelegate {
...
+        let supportVC = SupportTableViewController()
+        let navigationController = UINavigationController(rootViewController: supportVC)
+        topmostPresentedViewController.show(navigationController, sender: nil)
+    }
+}
+
+// MARK: - Routing
+
+private extension DomainSelectionViewController {
+    private func pushRegisterDomainDetailsViewController() {
```
Comment (staskus) [inline WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift:847]: Copied from `RegisterDomainSuggestionsViewController`

File: WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift [hunk 6]
```diff
@@ -715,3 +844,198 @@ extension WebAddressWizardContent: UITableViewDelegate {
...
+            }
+
+            WPAnalytics.track(.enhancedSiteCreationDomainsSelected, withProperties: domainSuggestionProperties)
+        default:
+            let properties: [AnyHashable: Any] = ["domain_name": domainSuggestion.domainName]
+            self.track(.domainsSearchSelectDomainTapped, properties: properties, blog: coordinator?.site)
+        }
+    }
+
+    func track(_ event: WPAnalyticsEvent, properties: [AnyHashable: Any] = [:], blog: Blog? = nil) {
```
Comment (staskus) [inline WordPress/Classes/ViewRelated/Domains/DomainSelectionViewController.swift:952]: Copied from `RegisterDomainSuggestionsViewController`

**Review by guarani — COMMENTED** [review]
Hey @staskus, thanks for tackling this refactor! 🏅 I took a first pass at reviewing and focused on testing today without looking much at the code. I split testing into two scenarios:
- user on the Free plan
- user on paid annual plans who already have a domain

I haven't tested users on the Free plan who already have a domain, users on paid plans who haven't claimed their free domain for a year yet, or users on monthly paid plans (and so are not eligible for a free domain). I also haven't finished a checkout to actually confirm the whole flow yet.

---

### Free plan scenarios

I created a new account and a new site
- [x] Site creation with free WP.com subdomain

I tested these flows up until the checkout:
- [x] Free to Paid Dashboard Card
- [ ] Register Dashboard Card
- [x] All Domains (Just buy a domain)
- [x] All Domains (Existing WP.com site)
- [x] Site Domains

> Register Dashboard Card

Is this referring to this dashboard card (pcdRpT-21K-p2) we implemented to sell domains without a plan? If so, since it was removed in https://github.com/wordpress-mobile/WordPress-iOS/pull/21597, so I'm not sure why it's in the list of entry points (but I might be misunderstanding this).

[not a blocker, possibly out of scope] Users in the Dashboard -> More -> Domains flow are offered a domain for their existing site in this scenario.  Should we only offer domains with a paid plan purchase? The user has a site, so matching the above "Existing WordPress.com site" where a domain is offered with a plan (which brings benefits to their site) seems intuitive.

---

### Paid plan scenarios 

Note that here the user has a paid domain, so additional domains are either re-directs or not associated with any site.

I tested these flows up until the checkout:
- [x] Free to Paid Dashboard Card (not applicable)
- [ ] Register Dashboard Card (not sure if applicable; see above)
- All Domains (Just buy a domain): The checkout screen shows an odd error message: "I tried and failed to create products from signup"
- All Domains (Existing WP.com site): The checkout screen re-directs to the WP.com login screen 
- Site Domains: The checkout screen opens briefly and closes

|  |  |
| - | - |

**Comment by staskus** [comment]
Thanks for the review!

> Register Dashboard Card

I meant this card, sorry for not clarifying. It's shown when the user has a plan but hasn't claimed the domain.

image

> All Domains (Just buy a domain): The checkout screen shows an odd error message: "I tried and failed to create products from signup"
All Domains (Existing WP.com site): The checkout screen re-directs to the WP.com login screen
Site Domains: The checkout screen opens briefly and closes

Mmm.. Strange. It works for me. Were you proxied?

This refactoring only tackles the domain selection view. I didn't change any other flows so it shouldn't be affected.

> Users in the Dashboard -> More -> Domains flow are offered a domain for their existing site in this scenario. Should we only offer domains with a paid plan purchase? The user has a site, so matching the above "Existing WordPress.com site" where a domain is offered with a plan (which brings benefits to their site) seems intuitive.

If I understand the question correctly, this is what we're doing with https://github.com/wordpress-mobile/WordPress-iOS/pull/22261 project. Offering domains with a paid plan purchase, not only domains with redirects?

**Comment by staskus** [comment]
> I haven't tested users on the Free plan who already have a domain, users on paid plans who haven't claimed their free domain for a year yet, or users on monthly paid plans (and so are not eligible for a free domain)

### Free plan who already have a domain

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/92050de1-d7d1-4765-8ec6-b7be2611ff95

As I understand one of your previous questions was regarding this flow, when the user already has a domain. I agree we should always offer a chance to purchase a plan for such a user. This refactor is not changing any behavior previously developed. And https://github.com/wordpress-mobile/WordPress-iOS/pull/22261 is adding another entry point.

### Users on paid plans who haven't claimed their free domain for a year yet

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/61a620ab-9d4a-4779-85aa-1430096ca07c

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/0285cf01-a1b5-4cec-8b99-91c78a2ac221

I also noticed that "All Domains" flow is underdeveloped for this case, however, it's not related to any changes we're making at the moment:
https://github.com/wordpress-mobile/WordPress-iOS/issues/22267

### Users on monthly paid plans  (and so are not eligible for a free domain)

Thanks, I never tested this flow before now. For such users, we don't show any cards, because they have a plan but are not eligible for domains as you say.

I wonder if there's a way for us to prompt such users to switch from month plan to year plan and get a free domain 🤔 

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/6afe0682-8677-4155-b987-c55ddc9debbf

**Comment by staskus** [comment]
I fixed additional cases I caught now when testing. Not directly related to changes I made in this PR but necessary to make flows work better:

- Present domain flow modally from Registration Card (to be consistent with all the other flows) https://github.com/wordpress-mobile/WordPress-iOS/pull/22254/commits/354b030a0004048727db6cccab7bc0297e668da3
- Push checkout instead of presenting it, otherwise "success view" does not appear in some flows (https://github.com/wordpress-mobile/WordPress-iOS/pull/22254/commits/1b8ecaf1870718b8745903bc1d258349261bb81b)
- After I merged with trunk, colors stopped working in Domains view. Now they have to be accessed through Design System (https://github.com/wordpress-mobile/WordPress-iOS/pull/22254/commits/f8a7c317eb393934f5eded6bd01e9f42e301d22b)

**Comment by staskus** [comment]
@guarani, yes the plugins flow is as you described. It works.

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/853204e4-c3e7-445f-8cc1-9054b3d36fa8

**Comment by guarani** [comment]
Thanks for the replies @staskus! It seems to be working as expected. I reviewed the code as best I could, the changes are pretty big, so I tried to track most flows through the code instead of looking at all the changes one by one. 

#### Register Dashboard Card

> I meant this card, sorry for not clarifying. It's shown when the user has a plan but hasn't claimed the domain.

It makes total sense now. I'd forgotten about that card! I tested the flow to the "Register domain" screen, and it worked as expected.

> If I understand the question correctly, this is what we're doing with #22261 project. Offering domains with a paid plan purchase, not only domains with redirects?

You're right, I got mixed up there. 

#### All Domains

- [x] "Just buy a domain" flow
- [x] "Existing WordPress.com site" flow

#### Site Domains

This is working as expected.

### Other scenarios

> Mmm.. Strange. It works for me. Were you proxied?

I probably was proxied, but I re-tested proxied and un-proxied today, and it works now. I don't know what the issue was.

> I wonder if there's a way for us to prompt such users to switch from month plan to year plan and get a free domain 🤔

I think this should be possible and [upgrade credit will be applied](https://wordpress.com/support/manage-purchases/upgrade-your-plan/#:~:text=If%20your%20site%20is%20on%20a%20monthly%20plan%2C%20you%20can%20save%20money%20by%20choosing%20the%20annual%20or%20multi%2Dyear%20subscription%20when%20you%20check%20out.), so perhaps we only would need to add the prompt. It's not in scope here, though.

**Review by guarani — APPROVED** [review]

**Pull request merged**